src/servers/src/grpc/flight.rs

```diff
@@ -168,13 +195,155 @@ impl FlightCraft for GreptimeRequestHandler {
         );
         async {
             let output = self.handle_request(request, Default::default()).await?;
-            let stream: Pin<Box<dyn Stream<Item = Result<FlightData, Status>> + Send + Sync>> =
-                to_flight_data_stream(output, TracingContext::from_current_span());
+            let stream = to_flight_data_stream(output, TracingContext::from_current_span());
             Ok(Response::new(stream))
         }
         .trace(span)
         .await
     }
+
+    async fn do_put(
+        &self,
+        request: Request<Streaming<FlightData>>,
+    ) -> TonicResult<Response<TonicStream<PutResult>>> {
+        let (headers, _, stream) = request.into_parts();
+
+        let header = |key: &str| -> TonicResult<Option<&str>> {
+            let Some(v) = headers.get(key) else {
+                return Ok(None);
+            };
+            let Ok(v) = std::str::from_utf8(v.as_bytes()) else {
+                return Err(InvalidParameterSnafu {
+                    reason: "expect valid UTF-8 value",
+                }
+                .build()
+                .into());
+            };
+            Ok(Some(v))
+        };
+
+        let username_and_password = header(AUTHORIZATION_HEADER)?;
+        let schema = header(GREPTIME_DB_HEADER_NAME)?.unwrap_or(DEFAULT_SCHEMA_NAME);
```

Comment:
I believe it may not only be the 'schema' but also the 'dbname.' See: https://github.com/michaelscofield/greptimedb/blob/72257639803cdf3790e757a4b8fc3c5cf6e43dbd/src/common/catalog/src/lib.rs#L63